builtin/logical/database/path_creds_create.go

```diff
@@ -83,9 +87,8 @@ func (b *databaseBackend) pathCredsCreateRead() framework.OperationFunc {
 
 		// Create the user
 		username, password, err := db.CreateUser(ctx, role.Statements, usernameConfig, expiration)
-		// Unlock
-		unlockFunc()
 		if err != nil {
+			unlockFunc()
 			b.closeIfShutdown(role.DBName, err)
```

Comment:
rather than giving up the lock and requiring it in the `closeIfShutdown` function what if we unlocked after the `closeIfShutdown` call. That'll protect against another thread getting the lock before we reset the plugin.

Reply:
I originally went down that route but we don't always have the write lock when this method gets called.